src/Moq/Expressions/Visitors/ConstructorCallVisitor.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) 2007, Clarius Consulting, Manas Technology Solutions, InSTEDD.
+// All rights reserved. Licensed under the BSD 3-Clause License; see License.txt.
+
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+
+using Moq.Properties;
+
+namespace Moq.Expressions.Visitors
+{
+	internal class ConstructorCallVisitor : ExpressionVisitor
+	{
+		/// <summary>
+		/// Extracts the arguments from a lambda expression that calls a constructor.
+		/// </summary>
+		/// <param name="newExpression">The constructor expression.</param>
+		/// <returns>Extracted argument values.</returns>
+		public static object[] ExtractArgumentValues(LambdaExpression newExpression)
+		{
+			if (newExpression is null)
+			{
+				throw new ArgumentNullException(nameof(newExpression));
+			}
+
+			var visitor = new ConstructorCallVisitor();
+			visitor.Visit(newExpression);
+
+			if (visitor._constructor == null)
+			{
+				throw new NotSupportedException(Resources.NoConstructorCallFound);
+			}
+
+			return visitor._arguments;
+		}
+
+		private ConstructorInfo _constructor;
+		private object[] _arguments;
+
+		public override Expression Visit(Expression node)
+		{
+			switch (node)
+			{
+				case LambdaExpression _:
+				case NewExpression _:
+				case UnaryExpression unary when unary.NodeType == ExpressionType.Quote:
+					return base.Visit(node);
+				default:
+					throw new NotSupportedException(
+						string.Format(Resources.UnsupportedExpressionInConstructorCall, node.NodeType.ToString()));
```

Comment:
* In cases where a string resource contains placeholders (`{0}` etc.), the Moq codebase usually explicitly passes `CultureInfo.CurrentCulture` to `string.Format`. (I didn't come up with that one, but:) Please follow that convention for consistency.

* Instead of reporting just the node type, it would be better to report the whole sub-expression that wasn't understood (`node.ToStringFixed()`). You could even just reuse the preexisting resource `Resources.UnsupportedExpression`.

Reply:
(P.S., speaking of resources: You'll notice that you have a merge conflict in `Resources.resx`. I've removed a few superfluous string resources since you opened this commit. Please make sure you don't reintroduce those when rebasing & resolving the conflict.)